iroh-gateway/src/core.rs

```diff
@@ -474,82 +479,87 @@ async fn serve_fs(
     start_time: std::time::Instant,
 ) -> Result<GatewayResponse, GatewayError> {
     // FIXME: we currently only retrieve full cids
-    let (mut body, metadata) = state
+    let (body, metadata) = state
         .client
-        .get_file(
-            req.resolved_path.clone(),
-            &state.rpc_client,
-            start_time,
-            &state.metrics,
-        )
+        .get_file(req.resolved_path.clone(), start_time, &state.metrics)
         .await
         .map_err(|e| error(StatusCode::INTERNAL_SERVER_ERROR, &e, &state))?;
 
     add_ipfs_roots_headers(&mut headers, metadata.clone());
-    match metadata.unixfs_type {
-        Some(UnixfsType::Dir) => {
-            if let Some(dir_list_data) = body.data().await {
-                let dir_list = match dir_list_data {
-                    Ok(b) => b,
-                    Err(_) => {
-                        return Err(error(
-                            StatusCode::INTERNAL_SERVER_ERROR,
-                            "failed to read dir listing",
-                            &state,
-                        ));
-                    }
-                };
-                return serve_fs_dir(&dir_list, req, state, headers, start_time).await;
-            } else {
-                return Err(error(
-                    StatusCode::INTERNAL_SERVER_ERROR,
-                    "failed to read dir listing",
-                    &state,
-                ));
+    match body {
+        FileResult::Directory(res) => {
+            let dir_list: anyhow::Result<Vec<_>> = res
+                .unixfs_read_dir(
+                    &state.client.resolver,
+                    OutMetrics {
+                        metrics: state.metrics.clone(),
+                        start: start_time,
+                    },
+                )
+                .map_err(|e| error(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string(), &state))?
+                .expect("already known this is a directory")
```

Comment:
I don't really understand what this `expect` says.

Reply:
this unwraps the option, which is guranteed to be `Some` if it is a directory like structure

Reply:
maybe change to "This will always unwrap since we have already guaranteed it to be a directory"

or "Should never panic since it is guaranteed to be a directory"

Cause I agree, the phrasing is hard to parse